src/core/lib/security/authorization/matchers.h

```diff
@@ -0,0 +1,127 @@
+
+  StringMatcher(const StringMatcher& other);
+  StringMatcher& operator=(const StringMatcher& other);
+  bool operator==(const StringMatcher& other) const;
+
+  bool Match(absl::string_view value) const;
+
+  std::string ToString() const;
+
+  StringMatcherType type() const { return type_; }
+
+  // Valid for EXACT, PREFIX, SUFFIX and CONTAINS
+  const std::string& string_matcher() const { return string_matcher_; }
+
+  // Valid for SAFE_REGEX
+  RE2* regex_matcher() const { return regex_matcher_.get(); }
+
+  static_assert(static_cast<int>(StringMatcherType::EXACT) == 0 &&
+                    static_cast<int>(StringMatcherType::PREFIX) == 1 &&
+                    static_cast<int>(StringMatcherType::SUFFIX) == 2 &&
+                    static_cast<int>(StringMatcherType::SAFE_REGEX) == 3 &&
+                    static_cast<int>(StringMatcherType::CONTAINS) == 4,
+                "Wrong StringMatcherType index.");
+
+ private:
+  StringMatcherType type_ = StringMatcherType::EXACT;
+  std::string string_matcher_;
+  std::unique_ptr<RE2> regex_matcher_;
+  bool ignore_case_ = false;
+};
+
+class HeaderMatcher {
+ public:
+  enum class HeaderMatcherType {
+    EXACT,       // value stored in StringMatcher field
+    PREFIX,      // value stored in StringMatcher field
+    SUFFIX,      // value stored in StringMatcher field
+    SAFE_REGEX,  // value stored in StringMatcher field
+    CONTAINS,    // value stored in StringMatcher field
+    RANGE,       // uses range_start and range_end fields
+    PRESENT,     // uses present_match field
+  };
+
+  HeaderMatcher() = default;
+  HeaderMatcher(const std::string& name, HeaderMatcherType type,
+                const std::string& matcher, int64_t range_start = 0,
+                int64_t range_end = 0, bool present_match = false,
+                bool invert_match = false);
+
+  HeaderMatcher(const HeaderMatcher& other);
+  HeaderMatcher& operator=(const HeaderMatcher& other);
+  bool operator==(const HeaderMatcher& other) const;
+
+  HeaderMatcherType type() const { return type_; }
+  const std::string& name() const { return name_; }
+  // Valid for EXACT, PREFIX, SUFFIX, SAFE_REGEX and CONTAINS
+  const StringMatcher* matcher() const { return matcher_.get(); }
+
+  bool Match(const absl::optional<absl::string_view>& value) const;
+
+  std::string ToString() const;
+
+  static_assert(static_cast<int>(HeaderMatcherType::EXACT) == 0 &&
+                    static_cast<int>(HeaderMatcherType::PREFIX) == 1 &&
+                    static_cast<int>(HeaderMatcherType::SUFFIX) == 2 &&
+                    static_cast<int>(HeaderMatcherType::SAFE_REGEX) == 3 &&
+                    static_cast<int>(HeaderMatcherType::CONTAINS) == 4,
+                "Wrong HeaderMatcherType index.");
+
+ private:
+  std::string name_;
+  HeaderMatcherType type_ = HeaderMatcherType::EXACT;
+  std::unique_ptr<StringMatcher> matcher_ = nullptr;
+  int64_t range_start_;
+  int64_t range_end_;
+  bool present_match_;
```

Comment:
@markdroth 
Envoy only looks at the type, whether it is present_match specifier. It does not care about the value if it is true/false.
https://github.com/envoyproxy/envoy/blob/e80b8eec292382f28ac0428229ce165d12ddeb13/source/common/http/header_utility.cc#L136
xDS code considers the value. 
https://github.com/grpc/grpc/blob/91c4cbd51b5edf62a2612e857853891f8f817bd9/src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc#L427

I updated the code to use the value.  But shouldn't we follow Envoy behavior? WDYT?

Reply:
Thanks for pointing this out!

I'm actually very surprised that Envoy doesn't actually look at the value of this field.  This means that Envoy interprets this field the same way, regardless of whether the value is true or false.  @htuch, is this behavior intentional?

The comment in the proto file doesn't make this clear:

https://github.com/envoyproxy/envoy/blob/748a9009e840c4f8da3a4021aa33f81d68712d8f/api/envoy/config/route/v3/route_components.proto#L1857

The field is a boolean, so we assumed that if the value was false, that meant to match when the field is *not* present.  That's how we implemented it in C-core, and @voidzcy tells me that grpc-java did the same thing.  @menghanl can tell us how grpc-go works.

We should probably figure out how this is supposed to work and then change all of our implementations to do the same thing.

Reply:
After chatting with Harvey, I've filed https://github.com/envoyproxy/envoy/issues/14656 about this.

For now, I guess let's stick with the existing gRPC behavior, so that we don't break anything that people might be depending on now.